### PR TITLE
Find group chats by chat id

### LINKED ISSUE #31: Messages in Group Chats Not Sent
I am able to receive group messages but not reply. Single person to person messages work as expected. This occurred with two separate group messages with all users using imessage not mms. The below error shows up in Terminal. Thanks for any help you provide!

MacOS: 10.13.6

````
full applescript tell application "Messages"
    activate
    set budy0 to first buddy whose handle is "=2b<FIRSTNUMBER>"
        set buddy1 to first buddy whose handle is "=2b1<SECONDNUMBER>"
    set thisChat to make new text chat with properties {participants:{buddy0,buddy1}}
    send "<MESSAGE CONTENT> " to thisChat
  end tell
59:103: execution error: Messages got an error: Can't get buddy 1 whose handle =  "=2b<FIRSTNUMBER>". Invalid index. (-1719)
sendEvent of type m.room.message in !<ROOM>:website.net with m1234567890123.1
Event set to !<ROOM>:website.net with event id $<EVENTID>
`````

```
Error in handleMatrixEvent { Error: exited nonzero
    at ChildProcess. (/Users/USERNAME/matrix-puppet-imessage/src/imessage-send-group.js:49:18)
    at ChildProcess.emit (events.js:198:13)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:248:12)
  [stack]:
   'Error: exited nonzero\n    at ChildProcess. (/Users/USERNAME/matrix-puppet-imessage/src/imessage-send-group.js:49:18)\n    at ChildProcess.emit (events.js:198:13)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:248:12)',
  [message]: 'exited nonzero' } { age: 33,
  content:
   { body:
      'MESSAGE',
     msgtype: 'm.text' },
  event_id: '$kvKsRcxp8W44kwn_W5BwElx4F9fjPsDbkWkNH_aJLQ',
  origin_server_ts: 1111111111111,
  room_id: '!VEdCCGuqmEccwednBS:hostname.net',
  sender: '@<username>:hostname.net',
  type: 'm.room.message',
  unsigned: { age: 33 },
  user_id: '@<username>:hostname.net',
  getRoomId: { [Function] [length]: 0, [name]: '' },
  getId: { [Function] [length]: 0, [name]: '' } } ﻿
```

### PR DESCRIPTION
**Pull request by cvincent**
Fixes #31.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -144,7 +144,7 @@ class App extends MatrixPuppetBridgeBase {
       ghostIntent.client.sendReadReceipt (event);
       return this.receiptHistory.delete(roomId);      
     } catch (err) {
-      debug('could not send read event', err.message);
+      console.log('could not send read event', err.message);
```
Comment (nr23730) [inline index.js:147]: Is there any reason to put the error to the console instead of debug?

**Comment by nr23730** [comment]
@michaelnew Can you also have a look at this?

**Comment by michaelnew** [comment]
I think #31 happens because matrix-appservice-bridge escapes characters like `@` and `+` when setting user IDs in Matrix, but it doesn't un-escape them when we read the user ID back out and try to match it with an iMessage user. In the case of 1-to-1 messages we use the room name in Matrix (which is just a *not* escaped version of the user ID) to find the iMessage conversation, but for group messages we use the list of user IDs. So we end up looking for a group chat with usernames like `mikenew=40gmail.com`, which don't exist. 

Anyway, it looks like this avoids the issue by not relying on user IDs to match up group chats at all and just using the chat ID from the room name. Which is a way more sane way of doing it anyway. We should probably still try and get a fix in with matrix-appservice-bridge in case we want to match up user IDs in the future, but that's really beside the point.

I'm a big fan of this change. I've had issues (even with a hacky fix for the character escaping) of group messages forking into different matrix rooms because the list of users was out of order or who knows what. So I'm hoping this will end up fixing that too. I'm testing it right now, so far so good. 👍 from me.

File: index.js (hunk 1 above)
Comment (cvincent) [reply index.js:147]: I actually don't remember why I made this change. I think it had been polluting the logs because `debug` was undefined, but this was quite a while back before last night when I came back around to look into the group message issue.

**Comment by cvincent** [comment]
I believe the previous method was only creating `text chat`s, which are distinct from `chat`s. My first attempt to fix it was by massaging the handles it used to create the chat, and it "worked" as in it created a duplicate text chat with the same members, rather than using the already existing chat.

There are probably additional cases to handle. I didn't test if this works with chats that have at least one non-iMessage user (green text bubbles). One of the advantages to being my friend is occasionally receiving a random message reading "testing" followed by "wow it worked!!" but I didn't want to push it too far all at once, haha.

**Comment by michaelnew** [comment]
Ha, yeah my friends get those messages too :)

I didn't realize `text chat` was a thing, so maybe that's what I was seeing with the duplicate rooms. BTW if you poke around the [iOS headers](http://developer.limneos.net/index.php?ios=13.1.3&framework=IMCore.framework&header=IMChat.h) there's some funny stuff in there, like `isAppleChat`, `isBusinessChat`, and `isMakoChat`, and I have no idea what those are supposed to mean.

I had the same thought about SMS. I'm going to see if I can set up a group chat with an SMS user and test that out.

**Comment by michaelnew** [comment]
Actually scratch that. I don't think I can create a group chat with an SMS user since my iMessage account no longer has a phone number associated with it. So I won't be able to test that.

Regardless, I still think we should merge this change since group chats are currently broken. Even if there are edge cases that aren't covered here, it's better than what we have.

**Comment by cvincent** [comment]
I'd be willing to carve out some time to circle back and test out SMS stuff since I do have a working account with a phone number. Can't hard-commit on when, but *possibly* this week?

**Comment by ericmigi** [comment]
seeing this error when using the branch. the message goes through but this is reported in the puppet control room

```
Error in handleMatrixEvent Error: exited nonzero
at ChildProcess. (/Users/eric/dev/imessagefix/matrix-puppet-imessage/src/imessage-send-group.js:48:18)
at ChildProcess.emit (events.js:210:5)
at Process.ChildProcess._handle.onexit (internal/child_process.js:272:12) {
[stack]: 'Error: exited nonzero\n' +
' at ChildProcess. (/Users/eric/dev/imessagefix/matrix-puppet-imessage/src/imessage-send-group.js:48:18)\n' +
' at ChildProcess.emit (events.js:210:5)\n' +
' at Process.ChildProcess._handle.onexit (internal/child_process.js:272:12)',
[message]: 'exited nonzero'
} {
age: 82,
content: { body: 'ask me a question only Liz would know', msgtype: 'm.text' },
event_id: '$Hk9-4MJXmHsddi-mAT0U5hmHelCWKkrfklpht8vQiBE',
origin_server_ts: 1581299445632,
room_id: '!VsStpkBKNMJcVagdiA:eric.nova.chat',
sender: '@liz:eric.nova.chat',
type: 'm.room.message',
unsigned: { age: 82 },
user_id: '@liz:eric.nova.chat',
getRoomId: [Function] { [length]: 0, [name]: '' },
getId: [Function] { [length]: 0, [name]: '' }
} ﻿
```

File: index.js (hunk 1 above)
Comment (nr23730) [reply index.js:147]: Okay! Good point!

File: src/client.js [hunk 2]
```diff
@@ -23,8 +23,8 @@ class Client extends EventEmitter {
   sendMessage (id, service, text, file) {
     return iMessageSend(id, service != "iMessage" ? "sms" : "iMessage", text, file);
   }
-  sendGroupMessage (handles, text, file) {
-    return iMessageSendGroup(handles, text, file);
+  sendGroupMessage (id, text, file) {
+    return iMessageSendGroup(id, text, file);
```
Comment (nr23730) [inline src/client.js:27]: I think this requires also a change in https://github.com/matrix-hacks/matrix-puppet-imessage/blob/master/index.js#L196

This probably resolved the issue of  @ericmigi

**Review by nr23730 — CHANGES_REQUESTED** [review]

**Comment by nr23730** [comment]
> 
> 
> I think #31 happens because matrix-appservice-bridge escapes characters like `@` and `+` when setting user IDs in Matrix, but it doesn't un-escape them when we read the user ID back out and try to match it with an iMessage user. In the case of 1-to-1 messages we use the room name in Matrix (which is just a _not_ escaped version of the user ID) to find the iMessage conversation, but for group messages we use the list of user IDs. So we end up looking for a group chat with usernames like `mikenew=40gmail.com`, which don't exist.
> 
> Anyway, it looks like this avoids the issue by not relying on user IDs to match up group chats at all and just using the chat ID from the room name. Which is a way more sane way of doing it anyway. We should probably still try and get a fix in with matrix-appservice-bridge in case we want to match up user IDs in the future, but that's really beside the point.
> 
> I'm a big fan of this change. I've had issues (even with a hacky fix for the character escaping) of group messages forking into different matrix rooms because the list of users was out of order or who knows what. So I'm hoping this will end up fixing that too. I'm testing it right now, so far so good. 👍 from me.

Thank you for your reply! :) I'm going to merge if the issue of @ericmigi is resolved.

**Comment by rootkit007** [comment]
I have been running this branch for about a week with no issues. Everything works, including group messages
